pkg/estimator/server/replica/replica_test.go

```diff
@@ -0,0 +1,580 @@
+							Status:             corev1.ConditionFalse,
+							Reason:             corev1.PodReasonUnschedulable,
+							LastTransitionTime: metav1.Time{Time: now.Add(-2 * time.Minute)},
+						},
+					},
+				},
+			},
+			threshold: 5 * time.Minute,
+			expected:  false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result := podUnschedulable(tt.pod, tt.threshold)
+			assert.Equal(t, tt.expected, result)
+		})
+	}
+}
+
+// Mock implementations of the Kubernetes listers
+type mockPodLister struct {
+	pods map[string][]*corev1.Pod // namespace -> pods mapping
+}
+
+func (m *mockPodLister) List(selector labels.Selector) (ret []*corev1.Pod, err error) {
+	var pods []*corev1.Pod
+	for _, nsPods := range m.pods {
+		for _, pod := range nsPods {
+			if selector == nil || selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (m *mockPodLister) Pods(namespace string) listcorev1.PodNamespaceLister {
+	return &mockPodNamespaceLister{
+		pods: m.pods[namespace],
+	}
+}
+
+type mockPodNamespaceLister struct {
+	pods []*corev1.Pod
+}
+
+func (m *mockPodNamespaceLister) List(selector labels.Selector) (ret []*corev1.Pod, err error) {
+	var result []*corev1.Pod
+	for _, pod := range m.pods {
+		if selector == nil || selector.Matches(labels.Set(pod.Labels)) {
+			result = append(result, pod)
+		}
+	}
+	return result, nil
+}
+
+func (m *mockPodNamespaceLister) Get(name string) (*corev1.Pod, error) {
+	for _, pod := range m.pods {
+		if pod.Name == name {
+			return pod, nil
+		}
+	}
+	return nil, nil
+}
+
+type mockReplicaSetLister struct {
+	replicaSets map[string][]*appsv1.ReplicaSet // namespace -> replicasets mapping
+}
+
+func (m *mockReplicaSetLister) List(selector labels.Selector) (ret []*appsv1.ReplicaSet, err error) {
+	var replicaSets []*appsv1.ReplicaSet
+	for _, nsRS := range m.replicaSets {
+		for _, rs := range nsRS {
+			if selector == nil || selector.Matches(labels.Set(rs.Labels)) {
+				replicaSets = append(replicaSets, rs)
+			}
+		}
+	}
+	return replicaSets, nil
+}
+
+func (m *mockReplicaSetLister) GetPodReplicaSets(pod *corev1.Pod) ([]*appsv1.ReplicaSet, error) {
+	var matchingRS []*appsv1.ReplicaSet
+	for _, rsList := range m.replicaSets {
+		for _, rs := range rsList {
+			if rs.Namespace == pod.Namespace {
+				selector, err := metav1.LabelSelectorAsSelector(rs.Spec.Selector)
+				if err != nil {
+					continue
+				}
+				if selector.Matches(labels.Set(pod.Labels)) {
+					matchingRS = append(matchingRS, rs)
+				}
+			}
+		}
+	}
+	return matchingRS, nil
+}
+
+func (m *mockReplicaSetLister) ReplicaSets(namespace string) listappsv1.ReplicaSetNamespaceLister {
+	return &mockReplicaSetNamespaceLister{
+		replicaSets: m.replicaSets[namespace],
+	}
+}
+
+type mockReplicaSetNamespaceLister struct {
+	replicaSets []*appsv1.ReplicaSet
+}
+
+func (m *mockReplicaSetNamespaceLister) List(selector labels.Selector) (ret []*appsv1.ReplicaSet, err error) {
```

Comment:
This is a great mock function, and for these mocks, I'm wondering if in the future centralise them into a public test util file dedicated to providing mock staking capabilities externally.

Reply:
Yeah that makes sense, most mocks used in different packages are redundant. We can consider merging them in a single util file in the future.

Reply:
This would be a nice improvement to reflect our top-level design for testing, and to put parts of the framework into the public test package of the entire repository, looking forward to this pr.

Reply:
> This would be a nice improvement to reflect our top-level design for testing

agree

Reply:
We'll open a separate issue for it in future.